### ISSUE TITLE
[CB-12397] .gitignore instead of unwanted .npmignore after cordova create (7.0.1)

### ISSUE THREAD

**Issue by brodybits** [apache/cordova-discuss#69]
```
C:\Users\Chris\Documents\chris-work
λ cordova --version
7.0.1

C:\Users\Chris\Documents\chris-work
λ cordova create ctapp1
Creating a new cordova project.

C:\Users\Chris\Documents\chris-work

C:\Users\Chris\Documents\chris-work
λ ls -a ctapp1\
./  ../  .npmignore  config.xml  hooks/  package.json  platforms/  plugins/  res/  www/

C:\Users\Chris\Documents\chris-work
λ cat ctapp1\.npmignore
# OS X
.DS_Store

C:\Users\Chris\Documents\chris-work
λ
```

I would rather have `.gitignore` instead ref: [Apache CB-12397: fix .gitignore for plugins & platforms in app template](https://issues.apache.org/jira/browse/CB-12397)

**Comment by purplecabbage** [comment]
Yes, a npmignore does not make sense.  I think it should be a gitignore
However, I think that the choice of gitignore 'ing platforms and plugins is a documentation issue though.
Doing this for our users could cause more issues, and invariably when it comes time to publish to app stores, unfortunately there are usually still some changes required inside the platform projects themselves. 

I would make the same comment on the issue in jira, however it is impossible to load here in Amsterdam.  [CB-12397]

**Comment by brodybits** [comment]
Thanks @purplecabbage for your response about this issue and CB-12397. Yes I would agree that this is a documentation issue, and a major one on my part. I have also seen this issue with a free third-party template.

My one question is about Cordova beginners who may miss this kind of warning until it is "too late", especially in the case of following some of the third-party tutorials out there.

I do understand the problem with changes needed in the platform projects themselves.

**Comment by raphinesse** [comment]
@purplecabbage I see this quite the other way around. If everything works just fine, most people will not check the docs for a remark that tells them to ignore some extra folders.

However, If things don't work the way they should people will see what they can do about it. So they will patch the generated code in `platforms` and either remove `platforms` from `.gitignore` or `git add -f platforms`.

Please don't make it hard for people to do the right thing (i.e. *keeping generated files out of your VCS*).

**Comment by stevengill** [comment]
@raphinesse you've convinced me

**Comment by raphinesse** [comment]
OK, so IMHO an app created with the *default template* should have:

- No `.npmignore` (There seem to be no objections here)
- `.gitignore` with the following contents:

```
# macOS
.DS_Store

# Generated by package manager
node_modules/

# Generated by Cordova
/plugins/
/platforms/
```
`node_modules/` is created on the first `cordova platform add`.

And yes, I think we want all the slashes, since we don't want to ignore someone's `/www/js/plugins/` folder :sweat_smile:

Side note: I wonder how the `.npmignore` even made its way into the `cordova-app-hello-world` package. If I `npm pack` from master, there's no such file :thinking:

**Comment by brodybits** [comment]
Guys I think we also need apache/cordova-create#8 to include `.gitignore` in the generated project. As I said in apache/cordova-create#8 this is due to NPM behavior related to `.gitignore` & `.npmignore` discussed in npm/npm#1862, npm/npm#3763, and npm/npm#7252.

**Comment by raphinesse** [comment]
@brodybits Unfortunately you are right. I ran `npm pack` on the latest master of `cordova-app-hello-world` and `.gitignore` is missing (but no `.npmignore` either; that's good). I suggest we give it some other name like `gitignore` and rename it in `cordova-create` then :unamused:

**Comment by raphinesse** [comment]
Specifiying `template_src` in the `files` property of our `package.json` **will** include the `.gitignore`.
This would allow us to include it, w/out some special treatment in `cordova-create`.

We would then have to add any of the following files that we want to keep in the package to the "files" array too
```
CONTRIBUTING.md
RELEASENOTES.md
.github/PULL_REQUEST_TEMPLATE.md
``

In case someone wondered: Adding an empty `.npmignore` will *not* make npm include the `.gitignore`

**Comment by brodybits** [comment]
Agreed with @raphinesse, just closed the workaround solution I had proposed in apache/cordova-create#8.

**Comment by brodybits** [comment]
As I said in <https://github.com/apache/cordova-app-hello-world/pull/22#issuecomment-390277676> I think we still need a workaround solution like what I proposed in apache/cordova-create#8 due to existing NPM behavior, just reopened apache/cordova-create#8 for discussion. I would be happy to rebase and retest apache/cordova-create#8 sometime next week. I am about to start a long weekend, will be available on Tuesday.

**Comment by raphinesse** [comment]
To make a long story short: Even if you have `.gitignore` in your package, when installed with npm (via `cordova-fetch`) it will get renamed to `.npmignore`. This is to make it easy to check `node_modules` into git for deployed apps.

So we have to rename it back. And since we should need it only for templates, `cordova-create` should be the right place to do this.

We'd still need apache/cordova-app-hello-world/pull/22 to get the `.gitignore` into the package in the first place. Or we name it `_gitignore` in the templates. Both of which require documentation. :unamused: 

I'm not sure, but I think the `_gitignore` variant is easier to document. Any opinions on this?

**Comment by brodybits** [comment]
I would vote -1 for renaming `.gitignore` in templates. Ugly and confusing. Also possible risk of causing problems down the road, like what happened in NPM as discussed in npm/npm#1862 / npm/npm#3763 / npm/npm#7252.

Unfortunately I do not really understand what apache/cordova-app-hello-world#22 is doing but the change looks clean and simple. No objection on my part.

I am about to start a long weekend with limited email. Nice work @raphinesse!

**Comment by raphinesse** [comment]
@brodybits apache/cordova-app-hello-world#22 sets up a whitelist of files that should be included in the package. That seems to be tho only way to get a `.gitignore` into a package.

The problem w/ that is that it does not universally apply to all templates. People would have to check which contents they want. With renaming to `_gitignore`, you can give a short and clear instruction:

> If you want to include a `.gitignore` in your template, you have to name it `_gitignore`. It will be renamed automatically to `.gitignore` upon template expansion.

Don't get me wrong, I dislike the whole renaming thing. But it seems we won't get past it.

Another point worth considering: Having an actual `.gitignore` in their template might be problematic for the template author's usage of git.

**Comment by raphinesse** [comment]
I was wrong again. :embarrassed: We don't need apache/cordova-app-hello-world#22.

Given this, I agree with @brodybits here. Just name it .gitignore in the template and rename it back after installation.

The two disadvantages I outlined above won't really matter in the wild, I think.

**Comment by raphinesse** [comment]
Since I could not find npm's `.gitignore` shenanigans properly documented, I did some testing to be sure of it's behavior. For this reason, I ran the following command for various types of references to `cordova-app-hello-world` (using `npm@6.0.1`):

    npm install --production --no-save TEMPLATE_REF

This is also what `cordova-fetch` runs when invoked by `cordova-create`.

Without further ado, here are my results:

Template reference | Template versions | Fetched ignore file
---|---|---
`/foo/bar` or `file:///foo/bar` | master, 3.12.0 | `.gitignore`
`git+file:///foo/bar` | master, 3.12.0 | None
`/foo/bar.tgz` | master, 3.12.0 | None
`https://example.com/foo/bar.git` | master | None
`git+file:///foo/bar` | package-gitignore | `.npmignore`
npm package name | 3.12.0 | `.npmignore`

I think the two most important facts to notice are:

- The behavior of npm around `.gitignore` files is not consistent at all.
- There are reference types where we will not even get a `.npmignore` on the receiving end without something like apache/cordova-app-hello-world#22

Given these observations and the remarks from my previous comment here, I am strongly in favor of avoiding all of npm's `.gitignore` handling and just have people name their ignore file `_gitignore`.

What do you say, @brodybits?

**Comment by brodybits** [comment]
Agreed on my part (@raphinesse I just saw your comment). Aargh aargh.

I do think `_gitignore` not being an easily missed "hidden" file is a nice bonus.

**Comment by raphinesse** [comment]
Could you rebase and update apache/cordova-create#8 to rename `_gitignore` instead of `.npmignore`? That would be great!

Later today, I can file PRs for the remaining changes we need for this.

**Comment by brodybits** [comment]
@raphinesse I would be happy to do as you asked. A couple things:

I did apache/cordova-create#8 as if there would be .npmignore as created by the NPM scripts, hoping this will not be an issue anymore

I would like to replace shelljs calls with fs-extra like @dpogue did for cordova-common in apache/cordova-common#21. I would also rather do this in a new PR and then redo apache/cordova-create#8 based of that.

Please confirm if you think this is OK or not.

**Comment by raphinesse** [comment]
@brodybits 

Regarding migration to `fs-extra`: Like I wrote in https://github.com/apache/cordova-common/pull/21#issuecomment-390712055 I support that. If you feel it would be beneficial to do this before landing apache/cordova-create#8 then that's fine with me. If it doesn't matter, I'd try to land this improvement first.

Regarding `.npmignore` files: I would not give them any special treatment at all. If one is present, it will be copied like any other file. Either it's supposed to be there, or the template author did not pay attention to the docs. We cannot distinguish these cases, so I would not issue a warning.

**Comment by brodybits** [comment]
@raphinesse I just updated the title to cover the desired result.

I raised apache/cordova-create#14 to introduce use of `fs-extra`, remove use of shelljs from the spec, minimize use of shelljs in cordova-create index.js, and isolate the remaining usage. This is desired on my part since the changes I originally proposed in apache/cordova-create#8 are using `shell.cp` directly, which I think is not what we want in the end.

Unfortunately "rebasing" apache/cordova-create#8 will take some extra work since the spec has changed so much since I had done the work last year. I hope I can finish this later today or maybe tomorrow. It would definitely help me if someone can review and merge apache/cordova-create#14 (or implement another solution) within the next 1-2 days.

I agree with your comment regarding `.npmignore` files; I am hopeful that your idea of using `_gitignore` will mean that my change should not have to deal with `.npmignore` at all.
